sycl/include/sycl/ext/oneapi/experimental/group_helpers_sorters.hpp

```diff
@@ -111,6 +111,142 @@ template <typename Compare = std::less<>> class default_sorter {
+      : comp(comp_), scratch(scratch_) {}
+
+  template <typename Group, typename Ptr>
+  void operator()([[maybe_unused]] Group g, [[maybe_unused]] Ptr first,
+                  [[maybe_unused]] Ptr last) {
+#ifdef __SYCL_DEVICE_ONLY__
+    using T = typename sycl::detail::GetValueType<Ptr>::type;
+    if (last <= first)
+      return;
+
+    auto mem_req = memory_required<T>(Group::fence_scope, last - first);
+    // Per extension specification if scratch size is less than the value
+    // returned by memory_required, behavior is undefined.
+    if (scratch.size() >= mem_req) {
+      // Adjust the scratch pointer based on alignment of the type T.
+      void *scratch_ptr = scratch.data();
+      std::align(alignof(T), (last - first) * sizeof(T), scratch_ptr,
+                 /* space */ mem_req);
+      auto aligned_scratch_ptr = static_cast<std::byte *>(scratch_ptr);
+      sycl::detail::merge_sort(g, first, last - first, comp,
+                               aligned_scratch_ptr);
+    }
+#else
+    throw sycl::exception(
+        std::error_code(PI_ERROR_INVALID_DEVICE, sycl::sycl_category()),
+        "default_sorters::joint_sorter constructor is not supported on host "
+        "device.");
+#endif
+  }
+
+  template <typename T>
+  static constexpr size_t memory_required(sycl::memory_scope,
+                                          size_t range_size) {
+    return range_size * sizeof(T) + alignof(T);
+  }
+};
+
+template <typename T, typename CompareT = std::less<>,
+          std::size_t ElementsPerWorkItem = 1>
+class group_sorter {
+  CompareT comp;
+  sycl::span<std::byte> scratch;
+
+public:
+  template <std::size_t Extent>
+  group_sorter(sycl::span<std::byte, Extent> scratch_,
+               CompareT comp_ = CompareT{})
+      : comp(comp_), scratch(scratch_) {}
+
+  template <typename Group>
+  T operator()([[maybe_unused]] Group g, [[maybe_unused]] T val) {
+#ifdef __SYCL_DEVICE_ONLY__
+    auto range_size = g.get_local_range().size();
+    auto mem_req = memory_required(Group::fence_scope, range_size);
+    // Per extension specification if scratch size is less than the value
+    // returned by memory_required, behavior is undefined.
+    if (scratch.size() >= mem_req) {
+      std::size_t local_id = g.get_local_linear_id();
+      // Adjust the scratch pointer based on alignment of the type T.
+      void *scratch_ptr = scratch.data();
+      std::align(alignof(T), /* output storage and temporary storage */ 2 *
+                                 range_size * sizeof(T),
+                 scratch_ptr, /* space */ mem_req);
+      auto aligned_scratch_ptr = static_cast<std::byte *>(scratch_ptr);
+      T *local_copy = ::new (scratch.data() + local_id * sizeof(T)) T(val);
+      sycl::detail::merge_sort(g, reinterpret_cast<T *>(aligned_scratch_ptr),
+                               range_size, comp,
+                               aligned_scratch_ptr + range_size * sizeof(T));
+      val = *local_copy;
+    }
```

Comment:
Should we add something like `assert(false && "Scratch size less than value results in UB")` in the else part? It will be helpful in debugging.